### DISCUSSION TITLE
Execute JS in RequestHandler

### DISCUSSION THREAD

**Discussion by BeanYa** [cefsharp/CefSharp#3763] [General]
### CefSharp Version

88.2.90

### .Net Version

.Net Framework 4.6

### Description

I can execute JS code in FrameLoadEnd/FrameLoadStart/LoadingStateChanged by attach a async method

```
browser.FrameLoadEnd += async (s, e) =>{ await ... }
```

The page (a login page) I loaded use Vue.js and redirect by `e.$router.push('/xxx')`.

The FrameLoadEnd/FrameLoadStart/LoadingStateChanged event wasn't triggered after `e.$router.push('/xxx')`. called.

But I found request handler can intercept that on `OnBeforeBrowser`.

So I implemented a class inherited from `RequestHandler` and override `OnBeforeBrowser`.

I need to do something with the redirect url.

## What I trying to do.

In the override method `OnBeforeBrowser`, I execute a JS code `document.getElementById('xx')`, the task always show `WaitingForActivation`.

> `browser.CanExecuteJavascriptInMainFrame` is true before execute that.

And I tried add a 3s timeout to `ExecuteScriptAsync("document.getElementById('xx')", timeOut: TimeSpan.FromSeconds(3))`, the task was always canceled and throw a exception.

## Question

Is that I shouldn't execute JS code in the `RequestHandler`?

Is there any alternative solution to that?

**Comment by amaitland** [comment]
> Is that I shouldn't execute JS code in the RequestHandler?

If you are asking can you evaluate JavaScript in OnBeforeBrowse? Only if you were to cancel the navigation. 

> Is there any alternative solution to that?

Without knowing what `e.$router.push('/xxx')` actually does I cannot say.

**Reply by BeanYa** [reply]
Thanks.

**Reply by amaitland** [reply]
Is the website a single page application(SPA)? 

You can use DevTools to look at what's going on.

https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#javascript-debugging

**Reply by BeanYa** [reply]
Yes, it's a SPA built by [Vue.js](https://github.com/vuejs/vue).

It built by another team, and I need to login and get token from header.

And they don't exposed the API to us. 

I already debug it and found out the page will redirect after `e.$router.push('/xxx')`.

The old version of the login window was wrote by WebBrowser, the `Navigating` event was raised after `e.$router.push('/xxx')` call.

What I am doing is use CefSharp and Chromium to refactor that login window to improve performance.

I even tried `AddressChanged` event of `CefSharp`, but it still not been raised after the `e.$router.push('/xxx')` (Add a break point in DevTools).

Maybe I need to found another way to get that token or Debug that page again.

Thanks for your reply.

**Reply by amaitland** [reply]
If the token is actually a header then you can get it from the response.
